hkm/views/views.py

```diff
@@ -475,27 +475,61 @@ def get_template_names(self):
     def setup(self, request, *args, **kwargs):
         # Session expires after browser is closed
         request.session.set_expiry(0)
-        self.search_term = request.GET.get('search', '')
-        self.page = int(request.GET.get('page', 1))
+        self.url_params = {
+            "search": request.GET.get('search', ''),
+            "page": int(request.GET.get('page', 1)),
+            "date_from": request.GET.get('date_from', ''),
+            "date_to": request.GET.get('date_to', ''),
+            "author": filter(None, request.GET.getlist('author', None)),
+            "date": filter(None, request.GET.getlist('date', None))
+        }
         return True
 
     def handle_search(self, request, *args, **kwargs):
-        LOG.debug('Search', extra={'data': {'search_term': self.search_term, 'page': self.page}})
+        LOG.debug('Search', extra={'data': {'search_term': self.url_params['search'], 'page': self.url_params['page']}})
+
+        url_params = request.session.get('url_params', {})
+        search_term_changed = self.url_params['search'] != url_params.get('search', '')
+        page_changed = self.url_params['page'] != url_params.get('page')
+        author_changed = self.url_params['author'] != url_params.get('author')
+        date_changed = self.url_params['date'] != url_params.get('date')
+        date_from_changed = self.url_params['date_from'] != url_params.get('date_from')
+        date_to_changed = self.url_params['date_to'] != url_params.get('date_to')
+
+        session_facet_result = request.session.get('facet_result', None)
+
+        if not session_facet_result:
+            self.facet_result = self.get_facet_result(self.url_params['search'])
+            # date_from & date_to require full list of dates
+            self.all_dates = self.facet_result.get('facets', {}).get('main_date_str', [])
+        # If search term or year to year changed => fetch facets again
+        elif search_term_changed or date_from_changed or date_to_changed:
+            self.facet_result = self.get_facet_result(self.url_params['search'], self.url_params['date_from'], self.url_params['date_to'])
+            self.all_dates = request.session.get('all_dates')
+        else:
+            self.facet_result = request.session.get('facet_result')
+            self.all_dates = request.session.get('all_dates')
+
+        facets = {}
+        if self.url_params['author']:
+            facets['author_facet'] = self.url_params['author']
+        if self.url_params['date']:
+            facets['main_date_str'] = self.url_params['date']
+        if self.url_params['date_from'] or self.url_params['date_to']:
+            facets['search_daterange_mv'] = "[%s TO %s]" % (self.url_params['date_from'], self.url_params['date_to'])
```

Comment:
I wonder if we would need to do something like this here as well:

```suggestion
            date_from = self.url_params['date_from'] if self.url_params['date_from'] else "*"
            date_to = self.url_params['date_to'] if self.url_params['date_to'] else "*"
            facets['search_daterange_mv'] = "[%s TO %s]" % (date_from, date_to)
```